HistoricalDataFetcher.Classes/Utilities/ApiRequest.cs

```diff
@@ -0,0 +1,172 @@
+            var uri = new Uri(url);
+
+            using (var handler = new HttpClientHandler
+            {
+                ServerCertificateCustomValidationCallback = (sender, certificate, chain, sslPolicyErrors) => true,
+                AllowAutoRedirect = false
+            })
+            using (var httpClient = new HttpClient(handler))
+            {
+                httpClient.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", _accessTokenInstance.Access_Token);
+                var response = await httpClient.GetAsync(uri);
+                
+                if (response.StatusCode == HttpStatusCode.Moved || response.StatusCode == HttpStatusCode.Redirect)
+                {
+                    return await RunEndpointHelperAsync(response.Headers.Location.ToString(), acceptHeader);
+                }
+                else
+                {
+                    rawJsonToReturn = await response.Content.ReadAsStringAsync();
+                }
+            }
+
+            _cache.Add(url, rawJsonToReturn);
+            _cache.Persist();
+
+            return rawJsonToReturn ?? string.Empty;
+        }
+
+        /// <summary>
+        /// Authenticate access using the host location
+        /// </summary>
+        /// <returns>bool: True = success</returns>
+        private static async Task<bool> AuthenticateAsync()
+        {
+            var body = $"{{\"username\": \"{_username}\",\"password\": \"{_password}\"}}";
+
+            var uri = new Uri($"{UrlBase}/api/authentication/login");
+    
+            using (var handler = new HttpClientHandler
+            {
+                ServerCertificateCustomValidationCallback = (sender, certificate, chain, sslPolicyErrors) => true
+            })
+            using (var httpClient = new HttpClient(handler))
+            {
+                httpClient.DefaultRequestHeaders.Accept.Clear();
+
+                using (var content = new StringContent(body, Encoding.UTF8, "application/json"))
+                using (var response = await httpClient.PostAsync(uri, content))
+                {
+                    return SetAuthenticationToken(JsonConvert.DeserializeObject<AccessToken>(await response.Content.ReadAsStringAsync()));
+                }
+            }
+        }
+
+        /// <summary>
+        /// Set the authentication token for subsequent calls 
+        /// </summary>
+        /// <param name="token">Access token</param>
+        /// <returns>bool: True = success</returns>
+        private static bool SetAuthenticationToken(AccessToken token)
+        {
+            if (token?.Access_Token == null)
+            {
+                return false;
+            }
+
+            _accessTokenInstance = token;
+
+            //Have the interval expire the token 5 minutes before the token is set to expire to ensure there is some overlap
+            var interval = (_accessTokenInstance.Expires.AddMinutes(-5) - DateTime.Now).TotalMilliseconds;
```

Comment:
There's something wrong with this line. I always get a negative number here.

Reply:
You'll want to guard against negative numbers and default to like 55 seconds in the case that the user has 1 minute expiration. (Which I think is what's happening here, because I setup a user with that)

Also, instead of using DateTime.Now, you'll want to use the Date on the headers returned form the server.

See what I did in nodekit: 

https://github.com/metasys-server/nodekit/blob/63ede3e1268ed4f82276506ac37e6acd2debfe8d/index.js#L175-L184

https://github.com/metasys-server/nodekit/blob/63ede3e1268ed4f82276506ac37e6acd2debfe8d/index.js#L157-L163

Reply:
I've confirmed that with a 30 minute expiration time this works okay. But if they have a 5 minute expiration this will likely fail especially if you are using DateTime.Now instead of server time. In my testing I could be almost a minute off from the server.

Reply:
the _accessTokenInstance.Expires is the expiration time returned from server, deserialized into local time, so we can use DateTime.Now to do the calculation.

server returns expiration date as a DateTimOffset object, which contains timezone info.

if the minimum expiration time in 1 minute, we can update this function to refresh 30 seconds before expiration date.

Reply:
Mike's suggestion has been implemented.

`var interval = (_accessTokenInstance.Expires.AddSeconds(-30) - DateTime.Now).TotalMilliseconds;`